### ISSUE TITLE
[enhancement]: Premature Taproot-Asset Channel Exhaustion Due to Above-dust `lnd` Requirement

### ISSUE THREAD

**Issue by dstadulis** [lightninglabs/taproot-assets#888]
# Background
In scenarios where taproot assets are contained within bitcoin lightning channels,
LND has the requirement to push an above-dust amount of satoshis to the remote side of the channel to complete a HTLC payment.  

Potential complication with this LND requirement is: 
In the scenario a taproot-assets-enabled lightning channel contains a de minimis amount of satoshis, (in situ taproot-assets outputs have limits -- until #721 is fixed) the bitcoin satoshi-balance of the channel will be exhausted after a few (3-5?) payments (assuming a payment in the `to_local` direction doesn't occur).

# Deliverables
A method to ensure that a taproot-assets lightning channel can send and receive HTLCs without being constrained by the LND above-dust requirement.

# Proposed Remediation Options
* add an inbound fee discount of the incoming channel w/ a base fee the same as the asset carrier HTLC amt
* ?Alternatives

# Necessity
This functionality isn't required to demonstrate tapd payment-channel functionality but will be necessary to meet user-expectations before LL endorses Taproot-assets payment-channel functionality.

**Comment by jharveyb** [comment]
You can also set up a matching HTLC (with no anchored assets) in the opposite direction to net out the BTC amount transferred.

**Comment by Roasbeef** [comment]
> You can also set up a matching HTLC (with no anchored assets) in the opposite direction to net out the BTC amount transferred.

Yep, this was a very early idea for LN, but we just ended up doing the trimmed dust route instead. The bi-directional HTLC trick _does_ require one side to have some funds to begin with, but the `push_amt` feature does exist, so that could be utilized. Mechanically, I think we'd want to add a new special rule where we insert that HTLC in the other direction automatically (special evaluation rule).

**Comment by GeorgeTsagk** [comment]
> add an inbound fee discount of the incoming channel w/ a base fee the same as the asset carrier HTLC amt

I don't think this works well. It assumes there's some symmetry in inbound/outbound traffic, and if that was the case (assuming that remote side also uses non-dust htlcs) then a fix would not even be needed as the sats would auto-balance

>  I think we'd want to add a new special rule where we insert that HTLC in the other direction automatically (special evaluation rule).

given that the channel lifecycle is totally LND's job, then that would have to be exposed via LND API?
IIUC we want a special case where if we want to send just an above dust amount to the other side, the commitment will also add an incoming htlc of the same amt.

I think building this directly into lnd channels is way too involved for what we actually need.

We might be better of with an automation in litd/tapd that simply compensates any received dust? After all, it's in both parties' interest to keep the channel operational

**Comment by dstadulis** [comment]
## Remediation Options
1) A draft implementation on LND has been developed to achieve feature-bit-less HTLC symmetry in the channel state machine. While this design will be the ultimate solution, more comprehensive test coverage is needed to ensure no significant faults could occur.  E.g. Bugs in the channel state machine could result in loss of funds even outside the Taproot channel context.

2) A bodge: As an interim solution, induce, on invoice creation, litd to cause the receiver of the tap payment to pay the dust back to the sender. 

Notes:
#923 "Soft Cap": 100k sat channels as a default, which would be sufficient for making multiple payments. Users will need to rebalance after a fixed number of HTLCs. The advantages of this approach include no significant code changes and a clean codebase, while the drawbacks include potential UX issues.

**Comment by linden** [comment]
Hi @dstadulis @Roasbeef, we're working on adding non-custodial Lightning to our wallet & SDK at Joltz. we're looking at doing an LSP model (similar to Zeus wallet) but given the soft cap introduced by #923, this would require our LSP to allocate 100k sats for any asset channel opened to the users. this isn't very feasible at scale to have so much locked up funds just to appease the dust limit, we also don't really want to push many sats to our users in the case they force-close a channel.

looks like https://github.com/lightningnetwork/lnd/pull/9430 will largely fix this issue for us, curious on what the status is as it doesn't seem to have moved in 3 months - is there a target release for this and/or anything we're able to help with to get it pushed along?

our backup plan here is to fork and lower our soft-cap, then rebalance as needed.

**Comment by Roasbeef** [comment]
@linden I think maybe you were looking for this issue: https://github.com/lightninglabs/taproot-assets/issues/1254 ? 

That's a feature request to be able to tune the amt of sats put into the channel.

That `lnd` issue also gets at a different interaction: right now we always push a certain amount of sats to the other side with each payment, as zero value outputs don't propagate well, and we need to anchor the asset commitment to an actual transaction output. The fix implemented in that draft is to settle the HTLC, but don't push any _sats_ funds to the receiver, only asset related funds. 

Returning back to #1254, the current value was selected via a somewhat arbitrary heuristic. Before LN channels move to the new zero fee commitments, we need enough funds in the channel to ensure that we can pay enough fees to _get into_ the mempool. Once we're in the mempool, users can CPFP, assuming they have enough funds on-chain. The amount in the channel also needs to be high enough to pay for the fees incurred due to added HTLC outputs (each pending HTLC makes the commitment txn bigger). 

So it's "safe" to decrease the amt of sats we allocate to an asset channel, but only to a certain extent. If the amt is too low, then force closes are less reliable. FWIW, if these are end user unadvertised channels, then there's less risk as there're no forwarded HTLCs in both directions. So then the main constraint is: having enough sats in a channel to pay for the max number of pending HTLCs.

**Comment by ZZiigguurraatt** [comment]
Would like to clarify this was not completed, but we now have two more specific open issues instead:

1. https://github.com/lightninglabs/taproot-assets/issues/1305
2. https://github.com/lightninglabs/taproot-assets/issues/1254

**Comment by linden** [comment]
@Roasbeef had missed your reply, thanks for the context. lowering the sats amount is a good workaround. looks like #1305 is a more proper solution as we'd then only need enough sats to cover the in-flight HTLCs (+ force close) and moving assets will no longer mean moving sats? would significantly improve UX around channels with both assets and sats if so.

was this closed by mistake as neither of these sub-issues were completed yet?

**Comment by ZZiigguurraatt** [comment]
> [@Roasbeef](https://github.com/Roasbeef) had missed your reply, thanks for the context. lowering the sats amount is a good workaround. looks like [#1305](https://github.com/lightninglabs/taproot-assets/issues/1305) is a more proper solution as we'd then only need enough sats to cover the in-flight HTLCs (+ force close) and moving assets will no longer mean moving sats? would significantly improve UX around channels with both assets and sats if so.

I don't think we will be moving sats anymore, just need enough sats in the channel to handle any on chain fees that are required, which can be higher if doing a force close vs co-op close, or can vary based on the fee market and/or the number of in flight HTLC.

I think the way https://github.com/lightninglabs/taproot-assets/issues/1305 is going to work is:

1. HTLC goes in flight and attaches 354 sat.
2. HTLC goes out of flight and the new commitment sends the 354 sat back.
3. If the peer doesn't agree to send the 354 sat back, you can force close the channel and never use that peer again. If the peer wants to keep you happy, they will agree to send the 354 sat back when the HTLC comes out of flight.


> was this closed by mistake as neither of these sub-issues were completed yet?

We decided the sub issues are more specific so we decided to close this one and keep less issues open.



After https://github.com/lightninglabs/taproot-assets/issues/1305 is fixed, one may want https://github.com/lightninglabs/taproot-assets/issues/1254 fixed to lower the sats amount and only use the channel for assets **OR** increase the sats amount and use the channel for both sats or assets.

**Comment by linden** [comment]
@ZZiigguurraatt thanks, that makes sense. maybe incorrect but my read on `NoopAdd` was that the peer is only credited with the funds if their balance is below the dust limit, so the peer doesn't need to cooperate (as in send sats back).
